### ISSUE TITLE
Processing only few fields instead of the entire document

### ISSUE THREAD

**Issue by techjourney010** [spring-projects/spring-data-couchbase#1151]
Spring boot project generated with Spring initializr
Spring Boot version: 2.5.1-SNAPSHOT
Spring Data Couchbase:
![image](https://user-images.githubusercontent.com/82444414/122593052-c3356080-d06d-11eb-8253-9e0983dd55c3.png)

```
Couchbase configuration:
@Configuration
@EnableReactiveCouchbaseRepositories
public class CouchbaseConfiguration extends AbstractCouchbaseConfiguration{...}
```

**Issue**: I am able to execute CRUD operations on the documents, but **I cannot process only few fields instead of the full document**.

```
@Document
public class MyDocument {
   @Field
   private Double rest;

   @Field
   private LocalDateTime created;
   
   ... some other fields
}
```

**Using Projections** - `ClassCastException`
```
public interface IMyDocRest {
    Double getRest();
}
```
```
@Data
@AllArgsConstructor
public class MyDocRest {
    private Double rest;
}
```

```
@Repository
public interface MyDocumentRepository extends ReactiveCouchbaseRepository<MyDocument , String> {... the methods presented bellow ...}
```

`Flux<IMyDocRest> findByCreatedLessThan(LocalDateTime created);`

_Try with the interface projection:_

```
2021-06-18 15:01:52.597 DEBUG 20060 --- [ cb-io-http-2-1] s.d.c.r.q.ReactivePartTreeCouchbaseQuery : Created query  WHERE `created` < 2021-06-18T13:41:27.723 for * fields.
2021-06-18 15:01:52.632 ERROR 20060 --- [ cb-io-http-2-3] reactor.core.publisher.Operators         : Operator called default onErrorDropped

reactor.core.Exceptions$ErrorCallbackNotImplemented: java.lang.ClassCastException: class MyDocument cannot be cast to class IMyDocRest (MyDocument and IMyDocRest  are in unnamed module of loader 'app')
Caused by: java.lang.ClassCastException: class MyDocument cannot be cast to class IMyDocRest  (MyDocument  and IMyDocRest  are in unnamed module of loader 'app')
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:106) ~[reactor-core-3.4.6.jar:3.4.6]
	...
```

_Try with the class projection_ - the same error:

`java.lang.ClassCastException: class MyDocument cannot be cast to class MyDocRest (MyDocument and MyDocRest are in unnamed module of loader 'app') 
...`

**Querying** - `ParsingFailureException`

`import org.springframework.data.couchbase.repository.Query;`

```
@Query("SELECT rest FROM #{#n1ql.bucket} WHERE created < $1 AND #{#n1ql.filter}")
Flux<Double> findRestByCreatedLessThan(LocalDateTime created);
```
```
DEBUG 20012 --- [ cb-io-http-2-1] .c.r.q.ReactiveStringBasedCouchbaseQuery : Created query 
ERROR 20012 --- [ cb-io-http-2-1] reactor.core.publisher.Operators         : Operator called default onErrorDropped

reactor.core.Exceptions$ErrorCallbackNotImplemented: com.couchbase.client.core.error.ParsingFailureException: Parsing of the input failed {"completed":true,"coreId":"0x3f9bb7c300000001","errors":[{"code":3000,"message":"syntax error - at bucket"}],"idempotent":false,"lastDispatchedFrom":"127.0.0.1:53365","lastDispatchedTo":"127.0.0.1:8093","requestId":66,"requestType":"QueryRequest","retried":0,"service":{"operationId":"6e5d852d-e559-4cd8-94b0-756f8c2f4c25","statement":"SELECT rest FROM default-bucket WHERE created < $1 AND `_class` = \"com.myserv.model.domain.MyDocument \"","type":"query"},"timeoutMs":75000,"timings":{"dispatchMicros":7244,"totalDispatchMicros":7244,"totalMicros":32837}}
```
_Another try:_
```
@Query("#{#n1ql.selectEntity} where #{#n1ql.filter} AND created SATISFIES created <= $1 END")
Flux<MyDocument> findWithQueryByCreated(LocalDateTime localDateTime);
```
`reactor.core.Exceptions$ErrorCallbackNotImplemented: com.couchbase.client.core.error.ParsingFailureException: Parsing of the input failed {"completed":true,"coreId":"0x7f624d4500000001","errors":[{"code":3000,"message":"syntax error - at SATISFIES"}],"idempotent":false,"lastDispatchedFrom":"127.0.0.1:49514","lastDispatchedTo":"127.0.0.1:8093","requestId":83,"requestType":"QueryRequest","retried":0,"service":{"operationId":"dc7d2311-8292-4e89-988d-22ba0cc0bf62","statement":"SELECT META(`default-bucket`).id AS __id, META(`default-bucket`).cas AS __cas, `default-bucket`.* FROM `default-bucket` where `_class` = \"com.myserv.model.domain.MyDocument\" AND created SATISFIES created <= $1 END","type":"query"},"timeoutMs":75000,"timings":{"dispatchMicros":7099,"totalDispatchMicros":7099,"totalMicros":29366}}`

_Other queries, same parsing exception:_
`@Query("SELECT SUM(rest) AS x FROM #{#n1ql.bucket} where #{#n1ql.filter} AND created < $1")`
`@Query("SELECT META(`default-bucket`).id AS __id, META(`default-bucket`).cas AS __cas, SUM(rest) AS __restsum FROM #{#n1ql.bucket} WHERE #{#n1ql.filter} AND created < $1")`

_Others have faced these problems too:_
[example](https://forums.couchbase.com/t/couchbase-springdata-4-1-1-and-boot-2-4-query-error/29100)
[another example](https://daiji110.com/2020/03/14/spring-data-couchbase-aggregate-function/)
etc.

**[Spring Data Couchbase documentation](https://docs.spring.io/spring-data/couchbase/docs/current/reference/html/#couchbase.repository.n1ql)**
_"You can also do single projections in your N1QL queries (provided it selects only one field and returns only one result, usually an aggregation like COUNT, AVG, MAX…​). Such projection would have a simple return type like long, boolean or String. This is NOT intended for projections to DTOs."_ - But where are the examples?

I read 
`#{#n1ql.selectEntity} WHERE #{#n1ql.filter} AND test = $1`
is equivalent to
`SELECT #{#n1ql.fields} FROM #{#n1ql.bucket} WHERE #{#n1ql.filter} AND test = $1`

but I do not see an example like `SELECT fieldA, fieldB FROM #{#n1ql.bucket} WHERE #{#n1ql.filter} AND fieldC = $1` which would would be very helpful!

**Questions:**
1. How can I select, by querying only few fields from a document?
2. How can I use any kind of projections with Spring Data Couchbase?
3. How can I apply an aggregate function, like SUM, **on a field** from the document?

**Comment by mikereiche** [comment]
1) How can I select, by querying only few fields from a document?
On the template, using  as(<otherClass>)

List<UserBrief> found = couchbaseTemplate.findAll(User.class).as(UserBrief.class).all();

2) How can I use any kind of projections with Spring Data Couchbase?
By the above, or by using the project(String[] fieldNames) method on ById operations.

**Comment by mikereiche** [comment]
"but I do not see an example like SELECT fieldA, fieldB FROM #{#n1ql.bucket} WHERE #{#n1ql.filter} AND fieldC = $1"

Try it.  You will also need to project meta().id as __id, meta().cas as __cas.  And place back-ticks around n1ql.bucket in case it has special characters.

 SELECT fieldA, fieldB,  meta().id as __id, meta().cas as __cas FROM `#{#n1ql.bucket}`  WHERE #{#n1ql.filter} AND fieldC = $1"

**Comment by mikereiche** [comment]
Please reopen if the above solutions are not sufficient.

**Comment by mikereiche** [comment]
There was a bug where AbstractCouchbaseQuery was not applying the result type using .as() that is fixed in #1208 4.3.0-M3
see also #1213
